File: benchmarks/saxpy.c
// SAXPY-style microbenchmark — hand-written C side.
// Matches benchmarks/saxpy.fl line for line so the comparison is fair.
#include <stdio.h>

#define N 64

static double saxpy_sum(const double *x, const double *y) {
    double s = 0.0;
    for (int j = 0; j < N; ++j) s += 2.0 * x[j] + y[j];
    return s;
}

int main(void) {
    double x[N], y[N];
    for (int j = 0; j < N; ++j) { x[j] = 1.0; y[j] = 2.0; }
    double total = 0.0;
    for (int i = 0; i < 200000; ++i) {
        x[0] = total / 1.0e9;
        total += saxpy_sum(x, y);
    }
    printf("%g\n", total);
    return 0;
}

File: benchmarks/saxpy.py
"""SAXPY-style microbenchmark — NumPy side.

Matches benchmarks/saxpy.fl and benchmarks/saxpy.c: 64-element vectors,
200_000 iterations, the same `total / 1e9` perturbation on x[0].
"""
import numpy as np


def bench() -> float:
    x = np.ones(64)
    y = np.full(64, 2.0)
    total = 0.0
    for _ in range(200_000):
        x[0] = total / 1.0e9
        total += float(np.sum(2.0 * x + y))
    return total


if __name__ == "__main__":
    print(f"{bench():g}")
